zephyr/lib/userspace_helper.c

```diff
@@ -0,0 +1,177 @@
+	default:
+		return false;
+	}
+#endif /* DRV_HEAP_CACHED */
+}
+
+void *drv_heap_aligned_alloc(struct sys_heap *drv_heap, uint32_t flags,
+			     uint32_t caps, size_t bytes, int32_t align)
+{
+#ifdef DRV_HEAP_CACHED
+	const bool cached = (flags & SOF_MEM_FLAG_COHERENT) == 0;
+#endif /* DRV_HEAP_CACHED */
+
+	if (drv_heap) {
+#ifdef DRV_HEAP_CACHED
+		if (cached) {
+			/*
+			 * Zephyr sys_heap stores metadata at start of each
+			 * heap allocation. To ensure no allocated cached buffer
+			 * overlaps the same cacheline with the metadata chunk,
+			 * align both allocation start and size of allocation
+			 * to cacheline. As cached and non-cached allocations are
+			 * mixed, same rules need to be followed for both type of
+			 * allocations.
+			 */
+			align = MAX(PLATFORM_DCACHE_ALIGN, align);
+			bytes = ALIGN_UP(bytes, align);
+		}
+#endif /* DRV_HEAP_CACHED */
+		void *mem = sys_heap_aligned_alloc(drv_heap, align, bytes);
+#ifdef DRV_HEAP_CACHED
+		if (!cached)
+			return mem;
+		return sys_cache_cached_ptr_get(mem);
+#else	/* !DRV_HEAP_CACHED */
+		return mem;
+#endif	/* DRV_HEAP_CACHED */
+	} else
+		return rballoc_align(flags, caps, bytes, align);
+}
+
+void *drv_heap_rmalloc(struct sys_heap *drv_heap, enum mem_zone zone,
+		       uint32_t flags, uint32_t caps, size_t bytes)
+{
+	int32_t align = 0;
+
+	if (drv_heap) {
+#ifdef DRV_HEAP_CACHED
+		if (!zone_is_cached(zone))
+			flags |= SOF_MEM_FLAG_COHERENT;
+#endif /* DRV_HEAP_CACHED */
+		return drv_heap_aligned_alloc(drv_heap, flags, caps, bytes, 0);
+	} else
+		return rmalloc(zone, flags, caps, bytes);
+}
+
+void *drv_heap_rzalloc(struct sys_heap *drv_heap, enum mem_zone zone,
+		       uint32_t flags, uint32_t caps, size_t bytes)
+{
+	void *ptr;
+
+	ptr = drv_heap_rmalloc(drv_heap, zone, flags, caps, bytes);
+	if (ptr)
+		memset(ptr, 0, bytes);
+
+	return ptr;
+}
+
+void drv_heap_free(struct sys_heap *drv_heap, void *mem)
+{
+	if (drv_heap) {
+#ifdef DRV_HEAP_CACHED
+		void *mem_uncached;
+
+		if (is_cached(mem)) {
+			mem_uncached = sys_cache_uncached_ptr_get(
+				(__sparse_force void __sparse_cache *)mem);
+			sys_cache_data_flush_and_invd_range(mem,
+							    sys_heap_usable_size(drv_heap, mem_uncached));
+
+			mem = mem_uncached;
+		}
+#endif
+		sys_heap_free(drv_heap, sys_cache_uncached_ptr_get(mem));
+	} else {
+		rfree(mem);
+	}
+}
+
+void drv_heap_remove(struct sys_heap *drv_heap)
+{
+	if (drv_heap)
+		rfree(drv_heap->init_mem);
+	rfree(drv_heap);
+}
+
+#else /* CONFIG_USERSPACE */
+
+void *drv_heap_rmalloc(struct sys_heap *drv_heap, enum mem_zone zone,
+		   uint32_t flags, uint32_t caps, size_t bytes)
+{
+	return rmalloc(zone, flags, caps, bytes);
+}
+
+
+void *drv_heap_aligned_alloc(struct sys_heap *drv_heap, uint32_t flags, uint32_t caps,
+			     size_t bytes, int32_t align)
+{
+	return rballoc_align(flags, caps, bytes, align);
+}
+
+
+void *drv_heap_rzalloc(struct sys_heap *drv_heap, enum mem_zone zone, uint32_t flags, uint32_t caps,
+		       size_t bytes)
```

Comment:
hopefully we merge #10088 first, then we can merge this one with no zones or caps